### PR TITLE
Add pg_authid view

### PR DESCRIPTION
**Pull request by adriangb**
Add a view to replace `pg_catalog.pg_authid`

### CODE REVIEW

File: sql/views.sql [hunk 1]
```diff
@@ -326,3 +326,8 @@ WHERE
     OR pg_catalog.pg_has_role(current_user, owner, 'MEMBER') IS TRUE;
 
 GRANT SELECT ON ALL TABLES IN SCHEMA timescaledb_information TO PUBLIC;
+
+-- View for viewing non-superuser passwords
+CREATE OR REPLACE VIEW timescaledb_information.pg_authid
+WITH (security_barrier = true) AS
+SELECT * FROM pg_catalog.pg_authid WHERE rolsuper = 'false';
```
Comment (adriangb) [inline sql/views.sql:333]: Happy to make this something fancier, e.g. to exclude replication users, roles with expired passwords, etc.

**Comment by mkindahl** [comment]
> Add a view to replace `pg_catalog.pg_authid`

Hi @adriangb and thank you for the contribution!

What I do not understand is what this new view would give that `pg_roles` do not give. Is the purpose to have have role-based access to the rows of `pg_authid` and be able to read passwords?

**Comment by adriangb** [comment]
The password in `pg_role` is masked out. Try `SELECT rolpassword FROM pg_roles;`; This gives you access to the password for non-superusers. My primary use case is to use it with pgbouncer for `auth_query` passthrough auth but there may be others and there's no reason not to allow people to access their own passwords.

**Comment by adriangb** [comment]
I understand the example but maybe I'm missing what it proves. What I want is for `tsdbadmin` (or a pooler account specifically granted those permissions via a security definer function created by `tsdbadmin`) to be able to read password hashes.

Can't we make this view only accessible to `tsdbadmin` (by not doing `grant all on leaky to public ;` and instead doing `revoke all...` and `grant execute on leaky to tsdbadmin` or similar)?

**Comment by mkindahl** [comment]
> I understand the example but maybe I'm missing what it proves. What I want is for `tsdbadmin` (or a pooler account specifically granted those permissions via a security definer function created by `tsdbadmin`) to be able to read password hashes.
> 
> Can't we make this view only accessible to `tsdbadmin` (by not doing `grant all on leaky to public ;` and instead doing `revoke all...` and `grant execute on leaky to tsdbadmin` or similar)?

Yes, that is possible, but it requires additional measures in place. In essence, you need to make sure that the view only shows passwords that the `tsdbadmin` should see (and `tsdbadmin` is not a superuser).

Note that only users that belong to the `tsdbadmin` role and to roles created by the `tsdbadmin` role should be shown (the same applies to any role, so it is not specific to `tsdbadmin`). You can look at `timescaledb_information.log_errors` if you want some idea of how it can be implemented.

Also note that there are still risks involved with this since you are sending the password over the network so it is necessary to make sure that you have an encrypted connection to the database.

**Comment by adriangb** [comment]
> Yes, that is possible, but it requires additional measures in place. In essence, you need to make sure that the view only shows passwords that the tsdbadmin should see (and tsdbadmin is not a superuser).

Agreed. I'm not sure how you'd define that, I went with "any non-superuser".

> Also note that there are still risks involved with this since you are sending the password over the network so it is necessary to make sure that you have an encrypted connection to the database.

Of course. If you're setting up a connection pooler it's either this or you store the passwords in plaintext on the connection pooler, which seems much more unsafe and inflexible.

**Comment by mkindahl** [comment]
@adriangb We're happy to accept the contribution, but in order for this to not introduce a security issue, we would need the following changes:
- Make sure that for any user X, only rows belonging to role X is shown.
- Add tests to demonstrate that it works this way.

**Comment by adriangb** [comment]
> Make sure that for any user X, only rows belonging to role X is shown.

Do you mean that a user can only see their own password? The whole point is that you can have a user (lets say `tsdbadmin` by default) that can see all passwords (except superuser passwords)

**Comment by mkindahl** [comment]
> > Make sure that for any user X, only rows belonging to role X is shown.
> 
> Do you mean that a user can only see their own password? The whole point is that you can have a user (lets say `tsdbadmin` by default) that can see all passwords (except superuser passwords)

Well, no, of course not, but the user should not be allowed to see passwords for accounts that it is not explicitly granted permission to view. For example, if it is a non-superuser, it should not be allowed to see passwords of superusers.

If you have tests set up and try to do something similar to how it works for `timescaledb_information.job_errors` we can work from there and make sure that there are no security issues.

I realize that the "requirement" is a little vague, but I think you get the gist. We do not want to expose passwords or hashes to users that are not allowed to view them.

**Comment by feikesteenbergen** [comment]
This needs a bit more design, referring to a blog post here:

```
https://www.cybertec-postgresql.com/en/pgbouncer-authentication-made-easy
```

We could come up with a security definer functions that is limiting the allowed hashes. For example, something like:
```sql
CREATE FUNCTION public.get_password_hash (
   INOUT user     name,
   OUT   password text
) RETURNS record
   LANGUAGE sql SECURITY DEFINER SET search_path = pg_catalog AS
$$SELECT usename, passwd FROM pg_catalog.pg_shadow WHERE usename = p_user AND pg_catalog.pg_has_role(p_user, 'member') $$;
```

However, I'm not sure if we want to make this change to timescaledb. What problem are you solving exactly?

**Comment by adriangb** [comment]
> What problem are you solving exactly?

As per some of the comments above I was trying to set up pgbouncer myself (for reasons I don't think it's worth getting into) and to do that you either need to put all of the passwords in a text file and keep them updated or use `auth_query` which needs to be able to read the password hashes.

**Comment by feikesteenbergen** [comment]
@adriangb a security definer function (as above) runs in the security context of the user that created it (`postgres`). Therefore, access to `pg_authid` is possible *within* the function.

The usefulness of the function is that it clearly creates a boundary where we switch from a regular user to a superuser.

**Comment by adriangb** [comment]
> a security definer function (as above) runs in the security context of the user that created it (postgres), as opposed to the one running it

Yes I know that. But `postgres` is not available on Timescale cloud, only `tsdbadmin`. And `tsdbadmin` doesn't have access to `pg_authid`. So there is no user to define the function as!

**Comment by adriangb** [comment]
It's still not clear to me what the missing discussion or design is here. I just want to have `tsdbadmin` be a bit closer to the permissions of a superuser. This is a limitation of Timescale that doesn't seem to be necessary, at least according to [this discussion](https://youtu.be/IdS35eMChpM?si=d3LbVF2Ds9yn6NVT) and the fact that CrunchyBridge doesn't have this issue (they give you superuser access in their cloud offering).

**Pull request closed without merging**